### ISSUE TITLE
Do we need to revisit the pagination cursors?

### ISSUE THREAD

**Issue by hailwood** [nuwave/lighthouse#18]
I think we need to revisit the pagination and connections as a whole in lighthouse, there's been a few changes in Laravel 5.4 that makes things easier, such as the Relation becoming macroable so we're able to add the `toConnection()` macro to that as well.

Additionally there is now a new method on Eloquent - `forPageAfterId` which should work well as it's effectively cursor pagination in a nutshell.

Now the current cursor generation doesn't seem right to me, I've found things like if I ask for the first 1, and keep incrementing the cursor then I get items 1, 2, 3, 5 and it goes out of whack from there (note this is contiguous data, IDs 1-200 are there).

This seems to come from a disconnect between how we generate the cursor for pagination e.g. here https://github.com/nuwave/lighthouse/blob/master/src/Support/Definition/PageInfoType.php#L73

And how we decode the cursor e.g. here https://github.com/nuwave/lighthouse/blob/master/src/Support/Traits/GlobalIdTrait.php, which seem to expect just type:id causing the effect above.

I'm proposing that we standardize the cursors to be graphqlType:cursorColumn:cursorColumnValue

So the 5th item from a graphql type if `user` would look like `user:id:5`
This means that to get the items we can just do e.g.

```php
function resolve($args){
   return User::orderBy('created_at', 'desc')->toConnection($args);

   //or if we didn't want to alter the query...
   return User::getQuery()->toConnection($args);
}

//this is the macro definition
function toConnection($args){
    list($type, $column, $value) = $this->decodeCursor($args);
    $first  = isset($args['first']) ? $args['first'] : 15;
    $items = $this->forPageAfterId($first, $column, $value)->get();
    $total = $this->count();
    return new LengthAwarePaginator($items, $total, $first);
}
```

Thoughts?

**Comment by chrissm79** [comment]
@hailwood Love this! Didn't even notice the new `forPageAfterId` but it sounds like a pretty solid drop in replacement. I also really like the encode/decode functionality you're proposing here... so if someone wanted to base it off a different column (such as a timestamp) they could, we would just need to allow them to override the column name in their connection class.

I'll try to add in this functionality by next week unless you wanted to submit a PR. 

Thanks!!

**Comment by chrissm79** [comment]
@hailwood Actually, I don't think `forPageAfterId` would necessarily work. Primarily because it's not a `LengthAwarePaginator` so we don't have things like the total count. It also looks like we might run into issues when trying to sort `desc` or if we have a sort on a different field coming from an argument in the resolver. I'll look into the issues you're experiencing and see if I can find out what's going on, but in the meantime it seems like it would still be beneficial to move from a trait to a `Relationship` macro.

**Comment by hailwood** [comment]
Fixed via #65
